building Java csvSink

Nauna opened the pull request:
- This work allows the csvSink to be built and run on the server.
- Allows each property to have a javaToCSV/javaToCSVLabel  which is a string that will output specified formatting for the csv outputter(csvSink).
- clean up a few toCSV legacy formatting and overrides

goes hand in hand with this pr: https://github.com/nanoPayinc/NANOPAY/pull/7306

Nauna committed:
display model tableColumns if they exist

Nauna committed:
moving CSVSink to java implementation

Nauna committed:
defining axioms gathered for csvSink.prop

Nauna committed:
csvSink props list  being set to foam.core.Property

Nauna committed:
adding csvOutput - toCSV - to PropertyInfo

Nauna committed:
adjusting property overrides in csvSink

Nauna committed:
adjust put arrayType

Nauna committed:
stringBuilder, name change, working on making csvSink java

Nauna committed:
Merge branch 'master' into tableColumnsEQCSVoutput

Nauna committed:
Merge branch 'master' into tableColumnsEQCSVoutput

Nauna committed:
Merge commit 'b65517afb' into tableColumnsEQCSVoutput

Nauna committed:
Merge branch 'master' into tableColumnsEQCSVoutput

Nauna committed:
Merge commit '86877bf62a8cffaa78ec386dd93a4ed5a589d93a' into tableColumnsEQCSVoutput

Nauna committed:
csvSink to java progressions

Nauna committed:
of check in csvSink

Nauna committed:
adding javaGetter to props on csvSink

Nauna committed:
Merge commit '41f40b8197c7ef6bbb90921b8fd2ce54f9951ed2' into tableColumnsEQCSVoutput

Nauna committed:
adjusting types for javaGetter in csvSink

Nauna committed:
debugging csvSink

Nauna committed:
building Java of csvSink

Nauna committed:
adding more to the csvSink java implementation

Nauna committed:
implementing the javaCode in csvSInk

Nauna committed:
adjust property of csvSink

Nauna committed:
Merge commit 'fc8c6bbfc74eec52bc7003a53583e64080f5426e' into csvSinkToJava

Nauna committed:
change prop type csvSink.props

Nauna committed:
passing in tableColumns from unstyledTableView

Nauna committed:
setting ctrl prop filteredTableColumns in ApplicationController

Nauna committed:
changes toCSV and toCSVLabel on PropertyInfo

Nauna commented:
Current State of the world

![image](https://user-images.githubusercontent.com/10802216/60261693-841f0b00-98aa-11e9-9f46-a8bc10431302.png)

Nauna committed:
adding conext toCSV

Nauna committed:
properly giving propertyInfos toCSV override capabilities

Nauna committed:
merge with branch ReferencedPropertyOverrides

Nauna committed:
Merge branch 'master' into testcsvSinkToJava

Nauna committed:
optimizations

Nauna committed:
clean residual from csvSink rework

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -44,13 +49,33 @@ foam.CLASS({
+        this.output_(value);
+      },
+      javaCode: `
+        StringBuilder sb = new StringBuilder();
```
This isn't the right use of a StringBuilder. You'll want to create a property that you store the StringBuilder in and this method will be something like:
Under properties:
```
{
  name: 'sb',
  flags: ['java'],
  javaType: 'StringBuilder',
  javaFactory: 'return new StringBuilder();'
}
```
and this method turns into:
```
        if ( ! getIsNewLine() ) getSb().append(",");
        setIsNewLine(false);
        output_(value);
```
and then you can implement the sink's `eof` method in java:
```
{
  name: 'eof'
  javaCode: 'setCsv(getSb().toString());'
}
```
And in the other methods, don't create new string builders and just re-use getSb() to append.

Nauna replied:
nice!

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -76,34 +101,97 @@ foam.CLASS({
+        sb.append(getCsv());
+        String s = value.toString();
+
+        if (s.indexOf("\\"") != -1 || s.indexOf(",") != -1) {
```
You might want to use logic similar to the code block above to keep js and java behavior more consistent with each other. You can't use an mmethod but you can use if/else similar to the java json outputter: https://github.com/foam-framework/foam2/blob/master/src/foam/lib/json/Outputter.java#L213

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -76,34 +101,97 @@ foam.CLASS({
-    function put(obj) {
-      if ( ! this.of ) this.of = obj.cls_;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((element) => {
```
Did you break this? Aren't props a StringArray now?

Nauna replied:
:( ya i did. Fixed with next push

mcarcaso commented on src/foam/java/PropertyInfo.js:
```diff
@@ -285,6 +287,20 @@ foam.CLASS({
+            name: 'toCSV',
+            visibility: 'public',
+            type: 'void',
+            args: [{ name: 'x', type: 'foam.core.X' }, { name: 'obj', type: 'Object' }, { name: 'outputter', type: 'foam.dao.CSVSink' }, { name: 'prop', type: 'Object' }],
```
Use multiple lines for this so it's easier to read.

mcarcaso commented on src/foam/java/PropertyInfo.js:
```diff
@@ -285,6 +287,20 @@ foam.CLASS({
+            name: 'toCSVLabel',
+            visibility: 'public',
+            type: 'void',
+            args: [{ name: 'outputter', type: 'foam.dao.CSVSink' }, { name: 'prop', type: 'Object' }],
```
Here too

mcarcaso commented on src/foam/nanos/controller/ApplicationController.js:
```diff
@@ -60,6 +60,7 @@ foam.CLASS({
     'appConfig',
     'as ctrl',
     'currentMenu',
+    'filteredTableColumns',
```
You don't want this on the ApplicationController. This ties the ApplicationController to one set of filteredTableColumns but there's no reason we can't render multiple table views at once. I think troy's suggestion was to put this on `foam.comics.DAOController`

Nauna committed:
optimization for csvSink in regards to StringBuilder

Nauna committed:
fixing broken put in js of csvSink

Nauna committed:
making code more legible - PropertyInfo

Nauna committed:
making java logic the same as js logic csvSink

Nauna committed:
merged to pegged dev foam

Nauna committed:
Merge branch 'storeChangesCSVSing' into csvSinkToJava

Nauna committed:
clean CSVDriver and CSVSink - minor adj

Nauna committed:
moving ctrl.filteredTableColumns to enable seperate columns per dao

Nauna committed:
setting eof in csvSink

Nauna committed:
make MDAO call eof()

Nauna committed:
fixing tests from csv changes

Nauna committed:
merge master

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
documenting

adamvy commented on src/foam/comics/DAOController.js:
```diff
@@ -253,16 +261,16 @@ foam.CLASS({
+          var link = document.createElement('a');
+          link.setAttribute('href', encodedUri);
+          link.setAttribute('download', 'data.csv');
+          document.body.appendChild(link);
```
We should be cleaning up this link after the fact as well

adamvy commented on src/foam/dao/CSVSink.js:
```diff
@@ -76,34 +106,99 @@ foam.CLASS({
+      name: 'newLine_',
+      code: function() {
+        this.csv += '\n';
+        this.isNewLine = true;
```
I personally would have called this isFirst, but that's not major

adamvy replied:
Or maybe isFirstColumn

adamvy commented on src/foam/dao/CSVSink.js:
```diff
@@ -76,34 +106,99 @@ foam.CLASS({
+        if ( ! this.of ) this.of = obj.cls_;
+        var element = undefined;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((name) => {
```
I would have done isFirstRow and put the header outputting into its own function.

`if ( this.isFirstRow ) this.outputHeaders`

Not a big deal

adamvy commented on src/foam/lib/csv/Outputter.java:
```diff
@@ -176,7 +176,7 @@ protected void outputFObject(FObject obj) {
     while ( i.hasNext() ) {
       PropertyInfo prop = (PropertyInfo) i.next();
-      prop.toCSV(this, prop.f(obj));
+      output(prop.f(obj));
```
Why this change?  shouldn't the property's toCSV still be respected if it wants to control how it outputs itself to CSV?

mcarcaso replied:
The behaviour of toCSV has changed and isn't really compatible with this Outputter anymore so I'm guessing the safest thing to do was this. If I'm reading this correctly, this roughly preserves the previous behaviour with the exception of Enums. IMO, we should be aiming to nuke this outputter completely as it's replaced by the CSVSink.

adamvy replied:
I see that toCSV now takes a CSVSink?  That doesn't sound right to me.  CSVSink should just internally use a CSVOutputter.  As it stands we're setting outselves up for confusion and possible bugs as people write javaToCSV for properties but then that doesn't actually get called when the object is output the CSV.

mcarcaso replied:
Yeah the CSVSink is basically an Outputter. Let's do this:
- Rename CSVSink to CSVOutputter
- Rename put to outputFObject (so it's a similar interface to json Outputter)
- Rename reset to flush
- Implement toString which returns csv in js and and sb.toString in java

Then implement a CSVSink which creates an instance of an outputter and on put, calls outputFObject and on eof, does a csv = outputter.toString()

That sound ok?

mcarcaso replied:
So CSVSink turns to something like this (note that there's no javaCode in here which you'll need to fill in):
```
foam.CLASS({
  package: 'foam.dao',
  name: 'CSVSink',
  extends: 'foam.dao.AbstractSink',

  documentation: 'Sink runs the csv outputter, and contains the resulting string in this.csv',

  properties: [
    {
      class: 'String',
      name: 'csv',
      view: 'foam.u2.tag.TextArea'
    },
    {
      class: 'Class',
      name: 'of',
      visibility: 'HIDDEN'
    },
    {
      name: 'props',
      visibility: 'HIDDEN'
    },
    {
      name: 'outputter',
      transient: true,
      factory: function(props) {
        return this.CSVOutputter.create({
          of: this.of,
          props: this.props
        })
      },
    }
  ],

  methods: [
    function put(obj) {
      this.outputter.outputFObject(obj);
    },

    function eof() {
      this.csv = this.outputter.toString();
    },

    function reset() {
      this.outputter.flush();
      this.csv = '';
    }
  ]
});
```

adamvy commented on src/foam/u2/view/UnstyledTableView.js:
```diff
@@ -33,7 +33,8 @@ foam.CLASS({
     'editRecord?',
-    'selection? as importSelection'
+    'selection? as importSelection',
+    'filteredTableColumns'
```
Make this an optional import

adamvy commented on src/foam/u2/view/UnstyledTableView.js:
```diff
@@ -53,21 +54,28 @@ foam.CLASS({
-
-          return c;
-        }).filter(function(c) { return c; });
+        this.filteredTableColumns = [];
```
Build the array first as a local variable, and then set the imported value afterwards.  This is for two reasons.  1) Anybody listening on updates to filteredTableColumns would first get a property change even as its set to an empty array. and then 2) property change events don't fire when an array's contents change.

adamvy replied:
Actually you'd be better off implementing this as a reaction.

```
foam.CLASS({
  name: 'UnstyledTableView',
...
  imports: [ 'filteredTableColumns?' ],
  reactions: [ '', 'propertyChange.columns_', 'onColumnsUpdate' ],
  listeners: [
    function onColumnsUpdate() {
      // First check if the optional import was provided.
      if ( this.filteredTableColumns$ ) {
        this.filteredTableColumns$ = this.columns_.map(p => p.name);
      }
    }
  ]
});
```

Then you don't need to modify the expression at all, and if somebody explicitly sets columns_ then filteredTableColumns will still be updated appropriately.

Nauna replied:
interesting

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
adjustments to filteredTableColumns in UnstyledTableView

Nauna committed:
clean

Nauna committed:
extracting csvOutputter from csvSink

Nauna committed:
corrections to var names related to CSVOutputter

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
explicit class call in CSVSink

Nauna committed:
fix filteredTableColumns passing

Nauna committed:
debug fixes to csvSink

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
removing old outputter property method

Nauna committed:
format DigWegAgent to work with new CSVOutputter

Nauna committed:
compile fixes

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +24,58 @@ foam.CLASS({
+        });
+      },
+      javaFactory: `
+        return new foam.lib.csv.CSVOutputter.Builder(getX()).setOf(getOf()).setProps(getProps()).build();
```
Style suggestion: If you're setting more than one field, break into multiple lines:
```
return new foam.lib.csv.CSVOutputter.Builder(getX())
  .setOf(getOf())
  .setProps(getProps())
  .build();
```
Is much easier to read and maintain.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+      class: 'String',
+      name: 'csv',
+      view: 'foam.u2.tag.TextArea',
+      documentation: 'Used for js version of this outputter.'
```
If this is for js only then give it `flags: ['js']` and it won't be generated for java.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+      factory: function() {
+        return this.of.getAxiomByName('tableColumns').columns;
+      },
+      visibility: 'HIDDEN'
```
I'm not sure you should hide any of these properties. If I'm viewing the CSVOutputter from a DetailView then it's probably because I'm debugging and I'll want to see a lot of these fields.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        for (String propName: tableColumnNames) {
```
you can just use `getProps()` here and not declare a tableColumnNames above.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        { type: 'FObject', name: 'obj' }
+      ],
+      code: function() {
+        var element;
```
Why is this declared up here? It's only used in the loop below.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+      code: function() {
+        var element;
+        this.props.forEach((name) => {
+          element = this.of.getAxiomByName(name);
```
Rename element to axiom or prop.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+      javaCode: 'return getSb().toString();'
+    },
+    {
+      name: 'headerOutput',
```
rename this to outputHeader

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        { type: 'FObject', name: 'obj' }
+      ],
+      code: function(obj) {
+        var element;
```
Don't declare up here

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        if ( this.isFirstRow ) this.headerOutput();
+
+        this.props.forEach((name) => {
+          element = this.of.getAxiomByName(name);
```
rename to axiom/prop

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        this.newLine_();
+      },
+      javaCode: `
+        Object propObj;
```
Don't declare any of these variables up here

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+
+        if ( getIsFirstRow() ) headerOutput(obj);
+
+        for (String propName : tableColumnNames) {
```
just use getProps()

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        if ( ! isPropertySet("of") || getOf() == null ) setOf(obj.getClassInfo());
```
Why do you need to `isPropertySet` check?

mcarcaso commented on src/foam/core/PropertyInfo.java:
```diff
@@ -43,7 +44,8 @@
   public Parser csvParser();
   public void toJSON(foam.lib.json.Outputter outputter, Object value);
-  public void toCSV(foam.lib.csv.Outputter outputter, Object value);
+  public void toCSV(X x, Object obj, foam.lib.csv.CSVOutputter outputter, Object value);
```
I think we can do this:
`public void toCSV(X x, FObject obj, foam.lib.csv.CSVOutputter outputter);`
And the toCSV method can figure out the value by doing an `Object value = f(obj);`
`PropertyInfos` implement `Expr` which has an `f` method on it that takes an object and gives you an object. In the case of a PropertyInfo, it takes an FObject and gives you the value of that property on the object.

mcarcaso commented on src/foam/core/PropertyInfo.java:
```diff
@@ -43,7 +44,8 @@
   public void toJSON(foam.lib.json.Outputter outputter, Object value);
-  public void toCSV(foam.lib.csv.Outputter outputter, Object value);
+  public void toCSV(X x, Object obj, foam.lib.csv.CSVOutputter outputter, Object value);
+  public void toCSVLabel(foam.lib.csv.CSVOutputter outputter, Object value);
```
I don't think you need the `value` in this method at all do you?

Nauna replied:
No we don't. Thx for the reminder

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        String[] tableColumnNames = getProps();
+
+        for (String propName: tableColumnNames) {
+          propObj = obj.getProperty(propName);
```
Since this code block is basically identical to the JS version, try to use the same variable names for the same object if you can. columnProp would be renamed to axiom or prop depending on what you renamed it to above.
propObj shouldn't be necessary.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,227 @@
+        if ( getIsFirstRow() ) headerOutput(obj);
+
+        for (String propName : tableColumnNames) {
+          propObj = obj.getProperty(propName);
```
Similar comment here as above. Ditch propObj and rename columnProp to axiom/prop.

mcarcaso commented on src/foam/lib/csv/Outputter.java:
```diff
@@ -176,7 +176,7 @@ protected void outputFObject(FObject obj) {
     while ( i.hasNext() ) {
       PropertyInfo prop = (PropertyInfo) i.next();
-      prop.toCSV(this, prop.f(obj));
+      output(prop.f(obj));
```
This change can change the behaviour of anything using this Outputter. Please make it a priority to remove this Outputter ASAP so nobody tries to fix it if/when something breaks. Anything using this outputter needs to be updated to use the new outputter.

mcarcaso commented on src/foam/nanos/export/CSVDriver.js:
```diff
@@ -10,30 +10,37 @@ foam.CLASS({
   methods: [
     function exportFObject(X, obj) {
-      return this.outputter.toCSV(obj);
+      // if obj coming in is not from a model with tableColumns,
```
I don't think this comment adds much value.

mcarcaso commented on src/foam/u2/view/UnstyledTableView.js:
```diff
@@ -53,21 +54,28 @@ foam.CLASS({
     },
     {
       name: 'columns_',
+      documentation: `Note: filteredTableColumns was a property set in DAOController
```
This shouldn't be intertwined within this expression like this. Let's use a reaction instead. Add this to the model:
```
reactions: [
  ['', 'propertyChange.columns_', 'updateFilteredTableColumns']
],
listeners: [
  {
    name: 'updateFilteredTableColumns',
    code: function() {
      this.filteredTableColumns = this.columns_.map(a => a.name);
    }
  }
]
```

Nauna replied:
considering this should have been done weeks ago. Are you ok with me making a ticket and adjusting this after.?

mcarcaso replied:
Why not just do it now? I pretty much gave you the code you need to use above. It might even be less work to just use that than to create a ticket and describe what needs to be done.

Nauna committed:
setting props exclusively in csvOutputter

Nauna committed:
compile fixes csvOutputter, csvSink, DigWebAgent

Nauna committed:
default adjustments to props and of on CSVSink

Nauna committed:
style changes: var names, defauls, synax. on CSVOutputter

Nauna committed:
Merge branch 'master' into csvSinkToJava

Nauna committed:
Merge branch 'outWithOldOutputter' into csvSinkToJava

Nauna committed:
remove legacy related to csvOutputter

Nauna committed:
changing toCSV signature

Nauna committed:
adding interface to csvOutputter, changing toCSVLabel signature, defaulting javaToCSVLabel on FObjectProperty, ensuring default settings

Nauna committed:
clean csvOutputter

Nauna committed:
fixing FObjectProperty refinements for csv outputting

Nauna commented:
Test Cases:
- Dig: calls CSVSink and CSVOutputter
(To Test) ```1)  http://localhost:8080/service/dig?dao=accountDAO&cmd=select&format=csv```
```2)  http://localhost:8080/service/dig?dao=invoiceDAO&cmd=select&format=csv```
(Note on Test 2) accountDAO call will check defaults but testing invoiceDAO tests nested FObject properties.
- EFTTest : uses CSVUtil
- Export from localhost -> admin -> services -> transactionDAO : uses ExportDriver/ExportModal
- Ablii -> payable -> export as CSV: uses CSVDriver

Nauna committed:
removing something not used

Nauna committed:
moving filteredTableColumns to a reaction: UnstyledTableView

Nauna committed:
change filteredTableColumns to a map to columns

Nauna committed:
Merge branch 'master' into csvSinkToJava

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
-        return of.getAxiomsByClass(foam.core.Property)
-          .filter( (p) => ! p.networkTransient );
+      factory: function(of) {
+        if ( ! of ) return;
```
Maybe return `[]`?

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
+      javaFactory: `
+        if ( ! isPropertySet("of") ) {
+          propsIsSet_ = false;
+          return null;
```
Same here. Rather than marking it as unset just return an empty array? I'm hesitant to mess with the propIsSet_ cuz that's managed by the model. Also, I could be wrong but I think this will get clobbered right after the factory runs and set to true.

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
+      class: 'FObjectProperty',
+      of: 'foam.lib.csv.CSVOutputter',
+      transient: true,
+      factory: function(of, props) {
```
These arguments won't actually be getting passed in since this is a factory. Ditch these from here and use this.of and this.props below

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
+      transient: true,
+      factory: function(of, props) {
+        var csvOutputter = foam.lib.csv.CSVOutputter.create();
+        if ( of ) {
```
Is there any harm in removing these checks and just creating the CSVOutputter like:
```
this.CSVOutputter.create({
  of: this.of,
  props: this.props
})
```
?

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
+      },
+      javaFactory: `
+      foam.lib.csv.CSVOutputter csvOutputter = new foam.lib.csv.CSVOutputter.Builder(getX()).build();
+      if ( isPropertySet("of") ) {
```
Do this without the conditional if you can too.
```
foam.lib.csv.CSVOutputter csvOutputter = new foam.lib.csv.CSVOutputter.Builder(getX())
  .setOf(getOf())
  .setProps(getProps())
  .build();
```

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -120,15 +130,15 @@ foam.CLASS({
       class: 'Function',
-      value: function(obj, outputter, prop) {
-        outputter.output(obj ? obj[prop.name] : null);
+      value: function(obj, outputter) {
```
x as first arg?

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -120,15 +130,15 @@ foam.CLASS({
       class: 'Function',
-      value: function(outputter, prop) {
-        outputter.output(prop.name);
+      value: function(x, outputter, obj) {
```
I don't think you need/want obj here. The code would be:
`outputter.outputValue(this.name);`
(where `this` is the foam.core.Property)

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
+        this.of.getAxiomsByClass(foam.core.Property)
           .forEach((axiom) => {
-            axiom.toCSV(obj ? obj[prop.name] : null, outputter, axiom);
+            axiom.toCSV(x, obj ? obj[this.name] : null, outputter, axiom);
```
I think you want:
`axiom.toCSV.call(axiom, x, obj && obj[this.name], outputter)`

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
       }
     },
+    {
+      name: 'javaToCSV',
```
Add these refinements in java/refinements.js

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
+      name: 'javaToCSV',
+      class: 'String',
+      value: `
+        if ( get(obj) == null && ! (get(obj) instanceof foam.core.FObject) ) {
```
This check isn't right. You should be doing the same check as js. If there's no `of` then you just output the object as is.

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
+          return;
+        }
+
+        java.util.List<foam.core.PropertyInfo> nestPropList = ((foam.core.FObject)get(obj)).getClassInfo().getAxiomsByClass(foam.core.PropertyInfo.class);
```
Here too, you want to base it off the `of` of the FObjectProperty. Using the ClassInfo of the property being stored can lead to bugs if it's a subclass of whatever the `of` is.

mcarcaso commented on src/foam/dao/CSVSink.js:
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
       name: 'toCSVLabel',
       class: 'Function',
-      value: function(outputter, prop) {
+      value: function(x, outputter, obj) {
```
Shouldn't need `obj` here. Base decisions on this.of

mcarcaso commented on src/foam/java/PropertyInfo.js:
```diff
@@ -285,6 +287,28 @@ foam.CLASS({
+            args: [
+              { name: 'x',          type: 'foam.core.X' },
+              { name: 'outputter',  type: 'foam.lib.csv.CSVOutputterInterface' },
+              { name: 'obj',        type: 'Object' }
```
You shouldn't need an obj here.

mcarcaso commented on src/foam/lib/csv/CSVOutputter.js:
```diff
@@ -0,0 +1,235 @@
+  name: 'CSVOutputter',
+
+  implements: [
+    'foam.lib.csv.CSVOutputterInterface'
```
I think the more common/consistent way to do this would be to have the interface called CSVOutputter and this model called CSVOutputterImpl

mcarcaso commented on src/foam/lib/csv/FObjectCSVOutputterDecorator.js:
```diff
@@ -0,0 +1,41 @@
+
+foam.CLASS({
+  package: 'foam.lib.csv',
+  name: 'FObjectCSVOutputterDecorator',
```
Create a ProxyCSVOutputter and make this extend that. This way, we can add methods to the interface and nothing else should need to get outputted.

mcarcaso replied:
Also rename this to PrefixedValueCSVOutputter ?

mcarcaso commented on src/foam/lib/csv/FObjectCSVOutputterDecorator.js:
```diff
@@ -0,0 +1,41 @@
+      of: 'foam.lib.csv.CSVOutputterInterface'
+    },
+    {
+      name: 'preLabelString',
```
Just call this `prefix`

mcarcaso commented on src/foam/lib/csv/Outputter.java:
```diff
@@ -1,235 +0,0 @@
-/**
- * @license
- * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Woohoo!

mcarcaso committed:
Merge branch 'master' into csvSinkToJava

mcarcaso committed:
- Give AbstractFObjectPropertyInfo a `of` method
- Fix up the method signatures for toCSV and toCSVLabel
- Clean up CSVSink and the refinements.
- Rename CSVOutputterInterface to CSVOutputter
- Rename CSVOutputter to CSVOutputterImpl
- Create a PrefixedCSVOutputter
- Fix CSVDriver when no filteredTableColumns are available.
- Check optional import before using it in UnstyledTableView

mcarcaso committed:
More cleanup

Nauna committed:
Merge pull request #1 from mcarcaso/csvSinkToJava

CSV Fixes

mcarcaso committed:
Skip unknown properties in java.

Nauna committed:
filtering propertyList to actually only be properties

mcarcaso reviewed (approved):
More featured and less code. Beautiful :)
